File: imogi_finance/imogi_finance/overrides/payment_request.py
from __future__ import annotations

import frappe
from erpnext.accounts.doctype.payment_request.payment_request import PaymentRequest

from imogi_finance.services.letter_template_service import render_payment_letter_html


class CustomPaymentRequest(PaymentRequest):
    def get_payment_letter_html(self):
        return render_payment_letter_html(self, letter_type="Payment Request Letter")


@frappe.whitelist()
def get_payment_request_payment_letter(name: str):
    doc = frappe.get_doc("Payment Request", name)
    return doc.get_payment_letter_html()

File: imogi_finance/imogi_finance/public/js/payment_request.js
frappe.ui.form.on('Payment Request', {
  refresh(frm) {
    maybeAddPaymentLetterButton(frm);
  },
});

function maybeAddPaymentLetterButton(frm) {
  if (frm.is_new() || frm.doc.docstatus !== 1) {
    return;
  }

  frm.add_custom_button(
    __('Payment Letter'),
    () => {
      frappe.call({
        method: 'imogi_finance.overrides.payment_request.get_payment_request_payment_letter',
        args: { name: frm.doc.name },
        callback(r) {
          if (!r.exc && r.message) {
            const w = window.open('', '_blank');
            w.document.write(r.message);
            w.document.close();
          }
        },
      });
    },
    __('Print'),
  );
}
